torch_ttnn/passes/lowering/to_tt_pass.py

```diff
@@ -499,6 +500,57 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
                 # Add additional logic to choose the appropriate memory_config type: DRAM or L1
                 return g.call_function(target_wrappers.clone, args=(args[0],))
 
+            # reference: tt-metal/tests/ttnn/operations/test_group_norm.py::test_group_norm_with_block_sharded_v2_8x8_grid_tile_layout
+            if node.target == torch.ops.aten.native_group_norm.default:
+                arg0_shape = get_shape(args[0])
+                if len(arg0_shape) != 4:
+                    return None
+                N, C, H, W = arg0_shape
+                input_tensor = args[0]
+                weight_tensor = args[1]
+                bias_tensor = args[2]
+                num_groups = args[6]
+                epsilon = args[7]
+                inplace = False
+                grid_size_x = device.compute_with_storage_grid_size().x
+                grid_size_y = device.compute_with_storage_grid_size().y
+                shard_shape = N * H * W // grid_size_x, C // grid_size_y
+                # input tensor
+                input_tensor_permute = g.call_function(ttnn.permute, args=(input_tensor, (0, 2, 3, 1)))
+                input_tensor_reshape = g.call_function(ttnn.reshape, args=(input_tensor_permute, (N, 1, W * H, C)))
+                # input mask
+                input_mask_tensor = g.call_function(
+                    ttnn.create_group_norm_input_mask, args=(C, num_groups, grid_size_y)
+                )
+                # gamma/beta
+                gamma = g.call_function(ttnn.create_group_norm_weight_bias_rm, args=(weight_tensor, C, grid_size_y))
+                beta = g.call_function(ttnn.create_group_norm_weight_bias_rm, args=(bias_tensor, C, grid_size_y))
+                output_tensor_l1 = g.call_function(
+                    target_wrappers.group_norm,
+                    args=(input_tensor_reshape,),
+                    kwargs={
+                        "input_mask": input_mask_tensor,
+                        "weight": gamma,
+                        "bias": beta,
+                        "num_groups": num_groups,
+                        "epsilon": epsilon,
+                        "inplace": inplace,
+                        "grid_size_x": grid_size_x,
+                        "grid_size_y": grid_size_y,
+                        "shard_shape": shard_shape,
+                    },
+                )
+
+                output_tensor_l1_reshape = g.call_function(ttnn.reshape, (output_tensor_l1, (N, H, W, C)))
+                output_tensor_l1_permute = g.call_function(ttnn.permute, (output_tensor_l1_reshape, (0, 3, 1, 2)))
+                new_node = output_tensor_l1_permute
+
+                node.replace_all_uses_with(new_node, delete_user_cb=lambda node: node != new_node)
+                node_users = list(new_node.users.keys())
+                for node_user in node_users:
+                    node_user.replace_all_uses_with(new_node)
```

Comment:
What is the reason we need to replace the users of users?

Reply:
I referenced from layer_norm

The output of `aten.native_group_norm.default` is (Y, mean, rstd), and the output of `ttnn.layer_norm` is Y, and the original garph is like this
```
group_norm = aten.group_norm
getitem = getitem(group_norm, 0)
return ((getitem,),)
```
and want the graph become this
```
group_norm = ttnn.group_norm
return (group_norm,)
```
so need to replace getitem's uses with group_norm

Reply:
I feel like the issue with this method is that it might go wrong if the graph isn't in the expceted form. For example if the graph calls `getitem(group_norm, 1)` to get mean, then we will mistakenly replace it with `Y` and results in wrong output.

We encountered a similar issue with batch_norm and max_pool2d and here is the solution we used:

https://github.com/tenstorrent/pytorch2.0_ttnn/blob/485dc5b6897faa35922dfb2949b356daf92d7da3/torch_ttnn/passes/lowering/to_tt_pass.py#L482-L491

It contains two parts:

1. `is_getitem_0_only_user` is to check if graph only calls `getitem(x, 0)`. So any unsupported `getitem(x, 1)` will fail-fast.
2. Instead of replacing `getitem(x, 0)` with `x`, it uses `g.call_function(target_wrappers.pack_to_tuple, (output,))` to pack `x` into `(x,)` so the `getitem` later will unpack it and get `x` correctly

IMO 1. is good to have to avoid troubles on debugging incorrect results later

Reply:
okok! This suggestion is great, thank for reminding, I've updated the code